include/kodlab_mjbots_sdk/lcm_subscriber.h

```diff
@@ -1,85 +1,149 @@
+
+  /**
+   * @brief Callback function for when an LCM message is received.
+   * @details This function serves as a callback for incoming LCM messages on
+   * `channel`.  It copies the decoded message data to `data`, and sets
+   * `new_message` to `true`.
+   * @param rbuf LCM recieve buffer containing raw bytes and timestamp
+   * @param channel channel name
+   * @param msg pointer to the incoming message data
+   */
+  void HandleMessage(const lcm::ReceiveBuffer *rbuf,
+                     const std::string &channel,
+                     const Message *msg)
+  {
+    mutex_.lock();
+    data = *msg;
+    new_message = true;
+    mutex_.unlock();
+  }
+};
+
+/**
+ * @brief LCM subscriber capable of subscribing to multiple channels
+ * @todo Implement subscription removal (i.e., unsubscribing)
+ */
+class LcmSubscriber : public AbstractRealtimeObject
+{
+public:
 
   /*!
-   * @brief subscribes to the lcm channel using handle message and handles ctrl c detection
+   * @brief Constructor an lcm subscriber
+   * @param realtime_priority realtime priority in range [1, 99]
+   * @param cpu cpu for this process
    */
-  void Run() override;
+  LcmSubscriber(int realtime_priority, int cpu)
+  {
+    cpu_ = cpu;
+    realtime_priority_ = realtime_priority;
+    Start();
+  }
 
-  std::string channel_name_;
+  /**
+   * @brief Add a new LCM channel to the the LCM object's subscriptions
+   * @tparam Message LCM message type for channel
+   * @param channel_name channel name
+   * @param handler `LcmMessageHandler`-inherited message handler object
+   * @return generated `lcm::Subscription` object
+   */
+  template<class Message>
+  lcm::Subscription *AddSubscription(std::string channel_name,
+                                     LcmMessageHandler<Message> &handler)
+  {
+    auto sub = lcm_.subscribe(channel_name,
+                              &LcmMessageHandler<Message>::HandleMessage,
+                              &handler);
+    subs_.emplace(channel_name, sub);
+    return sub;
+  }
+
+  /**
+   * @brief Accessor for `lcm::Subscription` objects
+   * @param channel_name channel name
+   * @return `lcm::Subscription` object on corresponding channel
+   */
+  [[nodiscard]] const lcm::Subscription *get_subscription(const std::string &channel_name) const
+  {
+    return subs_.at(channel_name);
+  }
+
+protected:
+
+  /**
+   * @brief LCM object
+   */
   lcm::LCM lcm_;
-};
 
-/************************************Implementation********************************************************************/
-template<class msg_type>
-void LcmSubscriber<msg_type>::Run() {
-  lcm_.subscribe(channel_name_, &LcmSubscriber::HandleMsg, this);
-  mutex_.unlock(); // Ensures mutex is unlocked
-  std::cout << "Subscribing to " << channel_name_ << std::endl;
-  while (!CTRL_C_DETECTED) {
-    lcm_.handleTimeout(1000);
+  /**
+   * @brief Map of channel names to corresponding `lcm::Subscription` objects
+   */
+  std::map<std::string, lcm::Subscription *> subs_;
```

Comment:
Presumably we will use this map to unsubscribe by name? Do the subscriptions do anything else?

Reply:
That was the intention, as well as accessing for customizing the individual `lcm::Subscription` objects.  These can be used to adjust message queue size if desired, as well as unsubscribe.